src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+
+            writer.WriteKey("specificDiagnosticOptions");
+            writer.WriteArrayStart();
+            foreach (var kvp in options.SpecificDiagnosticOptions)
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
+            }
+            writer.WriteArrayEnd();
+
+            if (options.Deterministic)
+            {
+                writer.Write("deterministic", true);
+                writer.WriteNull("localtime");
+            }
+            else
+            {
+                writer.Write("deterministic", false);
+                writer.Write("localtime", options.CurrentLocalTime.ToString(CultureInfo.InvariantCulture));
+            }
+
+            // Values which do not impact build success / failure
+            // - ConcurrentBuild
+            // - MetadataImportOptions:
+            // - Options.Features: deprecated
+            // 
+
+            // Not really options but they can impact compilation so we record the types. For the majority
+            // of compilations this is roughly the equivalent of recording the compiler version but it 
+            // could differ when customers host the compiler via the API.
+            writer.WriteKey("extensions");
+            writer.WriteObjectStart();
+
+            WriteType(writer, "syntaxTreeOptionsProvider", options.SyntaxTreeOptionsProvider?.GetType());
+            WriteType(writer, "metadataReferenceResolver", options.MetadataReferenceResolver?.GetType());
+            WriteType(writer, "xmlReferenceResolver", options.XmlReferenceResolver?.GetType());
+            WriteType(writer, "sourceReferenceResolver", options.SourceReferenceResolver?.GetType());
+            WriteType(writer, "strongNameProvider", options.StrongNameProvider?.GetType());
+            WriteType(writer, "assemblyIdentityComparer", options.AssemblyIdentityComparer?.GetType());
+            writer.WriteObjectEnd();
+        }
+
+        protected void WriteParseOptions(JsonWriter writer, ParseOptions parseOptions)
+        {
+            writer.WriteObjectStart();
+            WriteParseOptionsCore(writer, parseOptions);
+            writer.WriteObjectEnd();
+        }
+
+        protected virtual void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
+        {
+            writer.Write("kind", parseOptions.Kind);
+            writer.Write("specifiedKind", parseOptions.SpecifiedKind);
+            writer.Write("documentationMode", parseOptions.DocumentationMode);
+            writer.Write("language", parseOptions.Language);
+
+            var features = parseOptions.Features;
+            writer.WriteKey("features");
+            writer.WriteArrayStart();
+            foreach (var kvp in features)
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
```

Comment:
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.